### PR TITLE
sev-shim: use asm and naked_functions feature

### PR DESCRIPTION
**Pull request by haraldh**
also specify inline_asm for the x86_64 crate

### CODE REVIEW

**Comment by haraldh** [comment]
depends on https://github.com/enarx/enarx/pull/803

**Comment by npmccallum** [comment]
@haraldh This is a draft. Should it be promoted?

**Comment by haraldh** [comment]
> @haraldh This is a draft. Should it be promoted?

soon

**Comment by haraldh** [comment]
@npmccallum now it's ready... `wrfsbase` was merged upstream in `x86_64`

File: enarx-keep-sev-shim/src/asm.rs [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
+/// Fun read: http://www.rcollins.org/Productivity/TripleFault.html
+pub unsafe fn _enarx_asm_triple_fault() -> ! {
+    use x86_64::instructions::tables::lidt;
+    use x86_64::structures::DescriptorTablePointer;
+    // Create an invalid DescriptorTablePointer with no base and limit
+    let dtp = DescriptorTablePointer { limit: 0, base: 0 };
+    // Load the invalid IDT
+    lidt(&dtp);
+    // Provoke an #UD, which will lead to a triple fault, because of the invalid IDT
+    asm!("ud2", options(nomem, nostack));
```
Comment (npmccallum) [inline enarx-keep-sev-shim/src/asm.rs:18]: Can't we use `unreachable!()` instead here?

File: enarx-keep-sev-shim/src/usermode.rs [hunk 2]
```diff
@@ -2,12 +2,57 @@
...
+        # do a simulated return from interrupt
+        # this sets the segments and rip from the stack
+        iretq
+          ",
+    const USER_DATA_SEGMENT,
+    in(reg) sp,
+    const RFlags::INTERRUPT_FLAG.bits(),
+    const USER_CODE_SEGMENT,
+    in(reg) ip,
+    options(noreturn, nomem, att_syntax)
```
Comment (npmccallum) [inline enarx-keep-sev-shim/src/usermode.rs:56]: Intel syntax please. Let's follow Rust convention.

File: enarx-keep-sev-shim/src/syscall.rs [hunk 3]
```diff
@@ -12,11 +12,77 @@ use core::convert::TryFrom;
...
+    popq    %rdi
+
+    iretq
+          ",
+    const USR_RSP_OFF,
+    const KERNEL_RSP_OFF,
+    const USER_DATA_SEGMENT,
+    const USER_CODE_SEGMENT,
+    sym syscall_rust,
+    options(noreturn, att_syntax)
```
Comment (npmccallum) [inline enarx-keep-sev-shim/src/syscall.rs:84]: Intel syntax please. Follow Rust convention.

File: enarx-keep-sev-shim/src/asm.rs (hunk 1 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/asm.rs:18]: yeah, makes sense

File: enarx-keep-sev-shim/src/usermode.rs (hunk 2 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/usermode.rs:56]: waaaaah!!!

File: enarx-keep-sev-shim/src/syscall.rs (hunk 3 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/syscall.rs:84]: noooooo

File: enarx-keep-sev-shim/src/asm.rs (hunk 1 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/asm.rs:18]: hmm, no we can't ... unreachable!() calls panic before the `ud2`.

File: enarx-keep-sev-shim/src/usermode.rs (hunk 2 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/usermode.rs]: fixed

File: enarx-keep-sev-shim/src/syscall.rs (hunk 3 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/syscall.rs]: fixed

File: enarx-keep-sev-shim/src/asm.rs (hunk 1 above)
Comment (npmccallum) [reply enarx-keep-sev-shim/src/asm.rs:22]: ack

**Pull request merged**